### PR TITLE
Add support for decrypting ProofPoint URLdefense as well

### PR DESCRIPTION
**Pull request by nwhitehorn**
This adds support for unmangling URLs mangled by ProofPoint URLdefense instead of, or in addition to (don't ask -- this is a real thing), those mangled by SafeLinks. The decoder follows the documentation at https://help.proofpoint.com/Threat_Insight_Dashboard/Concepts/How_do_I_decode_a_rewritten_URL%3F

I've incidentally bumped the version number and marked this as compatible with Thunderbird 78.4+ (released today), which has backported messageDisplayScripts support.

### CODE REVIEW

File: change_message.js
Comment (cphyc) [inline change_message.js]: For the sake of consistency
```suggestion
let outlook_regex = new RegExp('/https://\w+\.safelinks\.protection\.outlook\.com/\?url=(\S+)\&amp;data[^\s"\']*', "g");
let proofpoint_regex = new RegExp('https://urldefense(?:\.proofpoint)?\.com/(v[0-9])/');
```

File: change_message.js
Comment (cphyc) [inline change_message.js]:
```suggestion
        let url = a.match(v2_pattern)[1].replace(/-/g, '%').replace(/_/g, '/');
```

File: change_message.js
Comment (cphyc) [inline change_message.js]:
```suggestion
        let encbytes = atob(url[2].replace(/_/g, '/').replace(/-/g, '+'));
        let encbytes_off = 0;
```

File: change_message.js
Comment (cphyc) [inline change_message.js]:
```suggestion
        outurl = url[1].replace(v3_token_pattern, (chunk) => {
            var len = 1;
            if (chunk.length > 1)
               len = length_codes.search(chunk[2]) + 2;
            var out = encbytes.substring(encbytes_off, encbytes_off + len);
            encbytes_off += len;
            return out;
        });
```

**Review by cphyc — COMMENTED** [review]
Thanks you for submitting a PR! I am overall happy with your suggested changes, provided you incorporate the suggested modifications (and assuming they work).

Could you also change the `Readme.md` to stipulate that it now also support ProofPoint urls?

File: change_message.js
Comment (cphyc) [inline change_message.js]:
```suggestion
        let url = a.match(v3_pattern);
```

**Review by cphyc — CHANGES_REQUESTED** [review]

**Comment by cphyc** [comment]
I have made some modifications to your branch, which I pushed to `https://github.com/cphyc/thunderbird_remove_safelinks/tree/proofpoint`.

**Comment by nwhitehorn** [comment]
Thanks for the fast reply! Would you like me to take care of the changes on my branch? Or were you going to (there seem to be some syntax errors etc. on yours that make it hard to test right now)?

**Comment by cphyc** [comment]
> Thanks for the fast reply! Would you like me to take care of the changes on my branch? Or were you going to (there seem to be some syntax errors etc. on yours that make it hard to test right now)?

I started working on it but did not manage to get a proper working version, so please update on your branch! I'll rebase on yours once merged.

Also, I have recently added tests (that I will merge later), could you please document here some known ProofPoint urls and their expected outcome. For ex. for outlook, I have
`https://na01.safelinks.protection.outlook.com/?url=http%3a%2f%2fwww.avanan.com` that should be be converted to `"http://www.avanan.com"`.

File: change_message.js
Comment (nwhitehorn) [reply change_message.js]: This one won't work entirely because encbytes_off needs to get updated.

Comment (cphyc) [reply change_message.js]: I guess it does
```javascript
> {
... let x = 0;
... var foo = () => {
..... ++x;
..... return x;
..... }
... }
undefined
> foo()
1
> foo()
2
> foo()
3
```

Comment (nwhitehorn) [reply change_message.js]: Huh, fascinating. Apparently my Javascript knowledge isn't great. Changed.

**Review by cphyc — APPROVED** [review]

**Comment by cphyc** [comment]
Once I have examples of mingled URLs and their decoded version (if possible for v1, v2 and v3 of ProofPoint), I'll hit the merge button!

**Comment by nwhitehorn** [comment]
I've added a commit with documentation updates. Here are some examples I either created by sending myself email or found on the internet (usually for older versions of ProofPoint's mangler, most of the v1 and v2 links I have no idea where they go or what they are):

URL: https://urldefense.com/v3/__http://www.nytimes.com__;!!HXCxUKc!gS3R_a6o2YLKi40IjjGycUqONlxHfNqrvHlhNGZwRBu9TKXx1tZpvsBf30hlJyo$
Unmangled URL: http://www.nytimes.com

URL: https://urldefense.proofpoint.com/v2/url?u=http-3A__www.google.com_&d=hdfskhafdslhkafdslhafdsflhfdsalhfadslhkfdsahjlfdsalhfdsalhfdsalhdfsalhfadslh
Unmangled URL: http://www.google.com/

URL: https://urldefense.proofpoint.com/v2/url?u=https-3A__api.onedrive.com_v1.0_shares_u-21aHR0cHM6Ly9vbmVkcml2ZS5saXZlLmNvbS9yZWRpcj9yZXNpZD1GRUQyQUZBQzMxMUQzMTFGITEyNzkmYXV0aEtleT0hQUg0c0JnQnNySXJNSkxJ_root_content&d=DwMCaQ&c=f-Ks-KiS2VRTRTUfkXMUvg&r=TQ7uKHB8pW4nr_C-ML5FaJJibvZVvm8gWKw_1RfXFus&m=ddb3WW62gJVoLm-9zi0G2Yoz0SJmYNCglDb0fg5C2pE&s=GWJ3mi6XJdAdcq9RyqlO5quzRLjOwtWij8MfOl9i2ec&e=
Unmangled URL: https://api.onedrive.com/v1.0/shares/u!aHR0cHM6Ly9vbmVkcml2ZS5saXZlLmNvbS9yZWRpcj9yZXNpZD1GRUQyQUZBQzMxMUQzMTFGITEyNzkmYXV0aEtleT0hQUg0c0JnQnNySXJNSkxJ/root/content&d=DwMCaQ

URL: https://urldefense.proofpoint.com/v2/url?u=http-3A__links.mkt3337.com_ctt-3Fkn-3D3-26ms-3DMzQ3OTg3MDQS1-26r-3DMzkxNzk3NDkwMDA0S0-26b-3D0-26j-3DMTMwMjA1ODYzNQS2-26mt-3D1-26rt-3D0&d=DwMFaQ&c=Vxt5e0Osvvt2gflwSlsJ5DmPGcPvTRKLJyp031rXjhg&r=MujLDFBJstxoxZI_GKbsW7wxGM7nnIK__qZvVy6j9Wc&m=QJGhloAyfD0UZ6n8r6y9dF-khNKqvRAIWDRU_K65xPI&s=ew-rOtBFjiX1Hgv71XQJ5BEgl9TPaoWRm_Xp9Nuo8bk&e=
Unmangled URL: http://links.mkt3337.com/ctt?kn=3&ms=MzQ3OTg3MDQS1&r=MzkxNzk3NDkwMDA0S0&b=0&j=MTMwMjA1ODYzNQS2&mt=1&rt=0&d=DwMFaQ

URL: https://urldefense.proofpoint.com/v1/url?u=http://www.bouncycastle.org/&k=oIvRg1%2BdGAgOoM1BIlLLqw%3D%3D%0A&r=IKM5u8%2B%2F%2Fi8EBhWOS%2BqGbTqCC%2BrMqWI%2FVfEAEsQO%2F0Y%3D%0A&m=Ww6iaHO73mDQpPQwOwfLfN8WMapqHyvtu8jM8SjqmVQ%3D%0A&s=d3583cfa53dade97025bc6274c6c8951dc29fe0f38830cf8e5a447723b9f1c9a
Unmangled URL: http://www.bouncycastle.org/

URL: https://urldefense.com/v3/__https://google.com:443/search?q=a*test&gs=ps__;Kw!-612Flbf0JvQ3kNJkRi5Jg!Ue6tQudNKaShHg93trcdjqDP8se2ySE65jyCIe2K1D_uNjZ1Lnf6YLQERujngZv9UWf66ujQIQ$
Unmangled URL: https://google.com:443/search?q=a+test&gs=ps

URL: https://www.nytimes.com
Unmangled URL: https://www.nytimes.com

URL: https://urldefense.com/v3/__http://example.com/index.php?title==**b__;0L_RgNCw0LLQvtCy0LDRjyvQt9Cw0YnQuNGC0LA!!HXCxUKc!gRk14VdM4bKmPfsqY4odp__dSc9upIBnlZA4wdaOmsnN-ipZLQHmW3osiBOeLj8$
Unmangled URL: http://example.com/index.php?title==Ð¿ÑÐ°Ð²Ð¾Ð²Ð°Ñ+Ð·Ð°ÑÐ¸ÑÐ°

**Comment by cphyc** [comment]
Awesome! Thanks for the contribution

**Comment by cphyc** [comment]
FYI, I've added tests for the URLs you provided except the last one, because of some weird UTF-8 issues in the testing suite
https://github.com/cphyc/thunderbird_remove_safelinks/blob/e3d041b26b7d24e869b9f855fe5aef9165a4eafb/test/tests.js#L36-L54

**Comment by nwhitehorn** [comment]
Thanks for merging! One note: I forget to edit the short description in package.json to note the new feature.

**Pull request merged**
